### PR TITLE
todoist-electron: 1.0.8 -> 8.3.3

### LINKED ISSUE #231111: Update request: todoist-electron 1.0.9 → 1.1.2
- Package name: todoist-electron
- Latest released version: 1.1.2
- Current version on the unstable channel: 1.0.9
- Current version on the stable/release channel: 1.0.9
- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

`todoist-electron` currently depends on `electron_21` which is EOL. I don't know how to determine from the appimage which version of Electron upstream is using.

**Notify maintainers**
@i077 @kylesferrazza

-----

Note for maintainers: Please tag this issue in your PR.

### PR DESCRIPTION
**Pull request by tebriel**
###### Description of changes

Update todoist-electron to the latest .AppImage and also update Electron to a supported version.

This closes #231111

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/todoist-electron/default.nix [hunk 1]
```diff
@@ -36,7 +37,7 @@ stdenv.mkDerivation rec {
   '';
 
   postFixup = ''
-    makeWrapper ${electron_21}/bin/electron $out/bin/${pname} \
+    makeWrapper ${electron_24}/bin/electron $out/bin/${pname} \
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/todoist-electron/default.nix:40]:
```suggestion
    makeWrapper ${electron_24}/bin/electron $out/bin/todoist-electron \
```

**Review by SuperSandro2000 — APPROVED** [review]

File: pkgs/applications/misc/todoist-electron/default.nix (hunk 1 above)
Comment (tebriel) [reply pkgs/applications/misc/todoist-electron/default.nix:40]: This already works and the variable was in place already, what's the need to replace this?

Comment (SuperSandro2000) [reply pkgs/applications/misc/todoist-electron/default.nix:40]: using pname wherever possible is an anti pattern and increases code complexity while also sneaking in potential errors with finalAttrs.

File: pkgs/applications/misc/todoist-electron/default.nix [hunk 2]
```diff
@@ -1,12 +1,13 @@
...
 
 stdenv.mkDerivation rec {
   pname = "todoist-electron";
-  version = "1.0.9";
+  version = "8.3.2";
 
   src = fetchurl {
-    url = "https://electron-dl.todoist.com/linux/Todoist-${version}.AppImage";
-    sha256 = "sha256-DfNFDiGYTFGetVRlAjpV/cdWcGzRDEGZjR0Dc9aAtXc=";
+    # https://electron-dl.todoist.com/linux/Todoist-linux-x86_64-8.3.2.AppImage
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/todoist-electron/default.nix:8]:
```suggestion
```
that comment is not helpful as in to discover where to download new versions or similar, we don't need to write out what the url would be with version substituted.

File: pkgs/applications/misc/todoist-electron/default.nix [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
+    url = "https://electron-dl.todoist.com/linux/Todoist-linux-x86_64-${version}.AppImage";
+    sha256 = "sha256-ASkjk/gTrwb7a99FcXwkue8Ds59Wvzi+YEmmjiGqv2o=";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/todoist-electron/default.nix:10]:
```suggestion
    hash = "sha256-ASkjk/gTrwb7a99FcXwkue8Ds59Wvzi+YEmmjiGqv2o=";
```
might as well

**Comment by kylesferrazza** [comment]
Hey @tebriel , have you tried logging in on your new build?
(See problems at https://github.com/NixOS/nixpkgs/pull/233514#issuecomment-1558377198)

**Comment by tebriel** [comment]
> Hey @tebriel , have you tried logging in on your new build?
> 
> (See problems at https://github.com/NixOS/nixpkgs/pull/233514#issuecomment-1558377198)

Hi thanks no, just re-using my existing session. I'll try that out.

**Comment by tebriel** [comment]
@kylesferrazza I used `sudo nixos-rebuild test -I nixpkgs=<path to your local nixpkgs folder> --fast` against this branch on a logged out Todoist install, I was presented with a link to "click to open a browser to log in" which successfully authenticated me. I also verified that I was running this updated version of todoist by looking at the running process (with `ps` output to see that it was indeed using the electron 24 I'd specified in this PR. I'm still very new to nix, do you mind checking out this branch and seeing if your problem persists?

**Comment by kylesferrazza** [comment]
> @kylesferrazza I used `sudo nixos-rebuild test -I nixpkgs=<path to your local nixpkgs folder> --fast` against this branch on a logged out Todoist install, I was presented with a link to "click to open a browser to log in" which successfully authenticated me. I also verified that I was running this updated version of todoist by looking at the running process (with `ps` output) to see that it was indeed using the electron 24 I'd specified in this PR. I'm still very new to nix, do you mind checking out this branch and seeing if your problem persists?

Absolutely will do, just away from my machine at the moment and wanted to make sure you tried this. 

> --edit I think it's also worth noting that I bumped the AppImage version of Todoist in this PR as well, maybe that fixes the electron compatibility issues you were experiencing?

Yeah, that makes sense, we can use `appimage-run` to run the new appimage and then check the electron version that upstream is using: https://github.com/NixOS/nixpkgs/issues/231111#issuecomment-1555904249

**Comment by kylesferrazza** [comment]
Running the latest appimage, I see:
```javascript
> navigator.appVersion
'5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Todoist/8.3.3 Chrome/112.0.5615.165 Electron/24.3.0 Safari/537.36 TodoistDesktop/8.3.3 TodoistElectronBridge/5'
```
so looks like we are good to move to Electron 24.

I confirmed I can build, run, and login to Todoist using this PR's branch.

Also, looks like the 8.3.3 AppImage is out now, if you want to update the version and hash now.

**Comment by kylesferrazza** [comment]
This closes #231111

**Review by kylesferrazza — CHANGES_REQUESTED** [review]
8.3.3 and new hash

**Comment by sersorrel** [comment]
this seems to work for me also :)

**Review by SuperSandro2000 — APPROVED** [review]

**Review by kylesferrazza — APPROVED** [review]

**Review by kylesferrazza — APPROVED** [review]

**Comment by tebriel** [comment]
Thanks everyone!

**Comment by jeffa5** [comment]
Would it be possible to backport this to the `nixos-23.05` branch please? @i077 @kylesferrazza

**Pull request merged**